prow/tide/status.go

```diff
@@ -360,19 +363,92 @@ func (sc *statusController) sync(pool map[string]PullRequest) {
 		tideMetrics.statusUpdateDuration.Set(duration.Seconds())
 	}()
 
-	// Query for PRs changed since the last time we successfully queried.
+	sc.setStatuses(sc.search(), pool)
+}
+
+func (sc *statusController) search() []PullRequest {
+	orgs, repos := sc.ca.Config().Tide.Queries.OrgsAndRepos()
+	freshOrgs, freshRepos := orgs.Difference(sc.trackedOrgs), repos.Difference(sc.trackedRepos)
+	oldOrgs, oldRepos := sc.trackedOrgs.Difference(orgs), sc.trackedRepos.Difference(repos)
+	// Stop tracking orgs and repos that aren't queried this loop.
+	sc.trackedOrgs.Delete(oldOrgs.UnsortedList()...)
+	sc.trackedRepos.Delete(oldRepos.UnsortedList()...)
+	// Determine the query for tracked PRs now before we modify 'trackedOrgs' and 'trackedRepos'.
+	var trackedQuery string
+	if sc.trackedOrgs.Len() > 0 || sc.trackedRepos.Len() > 0 {
+		trackedQuery = openPRsQuery(sc.trackedOrgs.UnsortedList(), sc.trackedRepos.UnsortedList())
+	}
+	queryStartTime := time.Now()
+
+	var lock sync.Mutex
+	var wg sync.WaitGroup
+	var allPRs []PullRequest
+	// Query fresh orgs and repos individually and since the beginning of time.
+	// These queries are larger and more likely to fail so we query for repos individually.
+	wg.Add(freshOrgs.Len())
+	for _, org := range freshOrgs.UnsortedList() {
+		go func(org string) {
+			defer wg.Done()
+
+			query := openPRsQuery([]string{org}, nil)
+			searcher := newSearchExecutor(context.Background(), sc.ghc, sc.logger, query)
+			prs, err := searcher.search()
+			if err != nil {
+				sc.logger.WithError(err).Errorf("Searching for open PRs in org:%s.", org)
+				return
+			}
+			lock.Lock()
```

Comment:
nit: this kind of locking is usually better when it's scoped to it's own function since defer is function scoped, if we have to add other logic later it will be nice to have the lock / unlock relevant bit explicitly scoped, we also repeat this logic further down

Reply:
I can reduce the repetition without making things too much more complicated, but I'm not sure what you are asking for in terms of scoping this to a function... The unlocking is already deferred and scoped to the anonymous function declared on line 390?